sway/desktop/render.c

```diff
@@ -451,22 +466,39 @@ static void render_surface_iterator(struct sway_output *output,
 
 		if (has_alpha) {
 			bool should_optimize_blur = view ? !container_is_floating(view->container) || config->blur_xray : false;
+
+			// Get a stencil of the window ignoring transparent regions
+			if (should_discard_transparent) {
+				fx_renderer_scissor(NULL);
+				fx_renderer_stencil_mask_init();
+
+				struct decoration_data stencil_deco_data;
+				memcpy(&stencil_deco_data, &deco_data, sizeof(struct decoration_data));
+				stencil_deco_data.discard_transparent = true;
+
+				render_texture(wlr_output, output_damage, &fx_texture, &src_box,
+						&dst_box, matrix, stencil_deco_data);
+
+				fx_renderer_stencil_mask_close(true);
+			}
+
```

Comment:
Apologies for the delay in reviewing, I really need to think about this bit. We could probably get rid of all of the new stencil functions by passing `discard_transparent` to the texture of the surface itself (instead of re rendering if its blurred), and then calling the blur rendering function from `fx_render_subtexture_with_matrix`. Not sure which solution is better. The suggestion here would be more portable for something like scenefx down the line

Reply:
Just seems wasteful to go through the texture twice

Reply:
> Just seems wasteful to go through the texture twice, afaik hyprland renders blur in the same function as the texture in their OpenGL code

Just confirmed that Hyprland does the same thing. I wonder if GLES2.0 supports blitting? In that case we could do that probably also remove our two framebuffer implementation too :)

Reply:
Yup you're correct, I had been hoping there was some gl magic we could do here but alas, GLES2 :/ okay! I'll give this a more thorough review tonight

Reply:
> Yup you're correct, I had been hoping there was some gl magic we could do here but alas, GLES2 :/ okay! I'll give this a more thorough review tonight

We can optimize it some more if we plan to add this into scenefx when using vulkan 😎

Reply:
For now, let's move the stencil functionality into `fx_render_blur`. This should clean up the stencil code quite a bit, and in the least reduce duplication in render.c

Reply:
> For now, let's move the stencil functionality into `fx_render_blur`

You mean `render_blur`? `fx_render_blur` is only used per up/downscale call